File: Code/C++/1964. Find the Longest Valid Obstacle Course at Each Position.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        const int n = obstacles.size();
        vector<int> res(n);
        vector<int> cur;
        for (int i=0; i<n; i++){
            int idx = upper_bound(cur.begin(), cur.end(), obstacles[i]) - cur.begin();
            if (idx == cur.size()){
                cur.push_back(obstacles[i]);
            }
            else{
                cur[idx] = obstacles[i];
            }
            res[i] = idx + 1;

        }
        return res;
    }

};

File: Code/Python/1964. Find the Longest Valid Obstacle Course at Each Position.py
from bisect import bisect_right

class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles):
        # Equivalently find the LIS (non-decreasing) at each pos
        res = []
        cur = []
        for x in obstacles:
            idx = bisect_right(cur, x) # the place it would insert
            if idx == len(cur):
                cur.append(idx)
            else:
                cur[idx] = x
            res.append(idx + 1)
        return res

